### ISSUE TITLE
Demucs models output near-silent stems on macOS 27 (MDXC models fine); --sample_rate 48000 shortens output

### ISSUE THREAD

**Issue by andriidohoda-ux** [ssmall256/mlx-audio-separator#4]
## Summary

On macOS 27 all Demucs models (`htdemucs`, `htdemucs_6s`) produce near-silent stems (around -57 dBFS RMS for a -17 dBFS mix). The residual (mix minus sum of stems) is essentially the whole mix. MDXC models work on the same machine (Roformer vocals, `MDX23C-DrumSep-aufr33-jarredou`), and so does the original PyTorch `demucs` 4.1.0 on MPS in the same venv, so the input file and environment look fine.

A second, smaller issue: `--sample_rate 48000` produces files that are too short (241.1 s instead of 262.4 s for a 48 kHz input). 262.44 × 44100 / 48000 = 241.1, so it looks like 44.1 kHz samples get written with a 48 kHz header.

## Environment

- Apple M4, macOS 27.0 (Darwin 27.0.0)
- Python 3.13.15 (Homebrew)
- mlx-audio-separator 0.1.7, mlx 0.31.2, mlx-metal 0.31.2, mlx-audio-io 1.3.11, mlx-spectro 0.7.0, numpy 2.5.3
- Demucs weights converted on first run into `~/.cache/demucs-mlx/*.safetensors` (torch 2.14.0, demucs 4.1.0)

## Repro

30 s stereo clip, 44.1 kHz float WAV, mean volume -18.2 dB:

```bash
mlx-audio-separator clip.wav -m htdemucs.yaml --output_format WAV
```

Mean volume per stem (ffmpeg `volumedetect`):

| Model | drums | bass | other | vocals |
|---|---|---|---|---|
| MLX `htdemucs` | -59.3 | -59.8 | -56.0 | -58.6 |
| PyTorch `demucs -n htdemucs -d mps` | -23.8 | -28.9 | -27.1 | -23.1 |

`htdemucs_6s` behaves the same way: every stem is between -58 and -60 dB. `mlx_audio_io.load` reads the input correctly (RMS -17.2 dB, peak 0.92), so the loss happens inside the Demucs path.

The log shows no warnings or errors. I tried these, and none of them changed the result:

- `MLX_AUDIO_SEPARATOR_FUSED_GROUPNORM_MODE=off`
- `MLX_AUDIO_SEPARATOR_DEMUCS_ISTFT_ALLOW_FUSED=0`
- `MLX_AUDIO_SEPARATOR_DETERMINISTIC_FUSED=1`

I couldn't test other MLX versions because mlx-audio-io pins `mlx==0.31.2`.

For comparison, on the same clip `model_bs_roformer_ep_317_sdr_12.9755.ckpt` gives Vocals -22.9 / Instrumental -20.7, and DrumSep on the PyTorch drums stem gives plausible per-drum levels.

I found this through [mlx-audio-separator-gui](https://github.com/fdebkowski/mlx-audio-separator-gui), which uses this engine.

**Comment by ssmall256** [comment]
Thanks for letting me know. I'll look into it tonight and try to push a new release in the next 12-24 hours.

**Comment by ssmall256** [comment]
Both issues are fixed in **0.1.8**, now on PyPI.

Thank you for the report — the level of detail in it is genuinely what made this
tractable. The dB measurements per stem, the `mix − Σstems ≈ mix` observation,
confirming MDXC/Roformer were fine on the same machine, checking PyTorch
`demucs` on MPS as a control, and listing every environment variable you had
already ruled out: that narrowed it to the Demucs weight-loading path before I
had run anything. Noting that you could not test other MLX versions because
mlx-audio-io pinned `mlx==0.31.2` turned out to matter too, and is also fixed.

**Near-silent stems.** The safetensors cache introduced in 0.1.7 was *written*
with MLX attribute names (`encoder.0.conv.weight`) but *read back* through the
PyTorch-style key walker used by the torch→MLX conversion path, which collapses
the `conv`/`layers` wrapper segments. The lookup had no `else` branch, so every
miss was silent: **360 of 573 tensors never matched and stayed at their random
initialisation.** A randomly-initialised U-Net emits noise-floor output, which is
exactly the −57 dBFS you measured, and identically for `htdemucs` and
`htdemucs_6s`. Every environment variable you tried is downstream of that, which
is why none of them changed anything, and MDXC/Roformer/VR use different loaders
so they were untouched.

Cache loading now validates keys and shapes against the constructed model and
raises on any mismatch instead of loading silently. **Existing caches do not need
to be regenerated** — the writer was always correct, only the reader was wrong.

**`--sample_rate` truncation.** Demucs runs at its trained 44.1 kHz, but stems
were written with the requested rate in the header and no resample in between,
so a 262.4 s track came out at 241.1 s — the 8.8% you measured. Stems are now
resampled to the requested rate before writing. The default 44.1 kHz path is
unchanged and does no extra work.

**MLX version flexibility.** mlx-audio-io 1.3.12 accepts `mlx>=0.31.2,<0.33` and
is verified against 0.32.2; Demucs output is bit-identical across both versions.
Note the extension is compiled at install time against whatever MLX pip resolves
in its isolated build environment, so if you pin an older MLX the loader will now
tell you clearly and print the exact commands to rebuild against it.

Also worth knowing if you are upgrading: Demucs defaults changed in 0.1.8 after
measurement. The fused GroupNorm/GLU Metal kernels are off by default (they cost
~20 dB SNR and were not faster), and the batch size default is 2 rather than 8
(roughly 2x faster and half the peak memory). Full details in the
[changelog](https://github.com/ssmall256/mlx-audio-separator/blob/main/CHANGELOG.md).

Please do reopen if anything still looks wrong on your setup.

**Comment by andriidohoda-ux** [comment]
Confirmed fixed on my machine (M4, macOS 27.0). Thank you for turning this around so quickly — and for the write-up. The random-initialisation explanation matches every symptom exactly, including why the stems sat at the noise floor no matter which environment variable I touched.

Verification on the same 30 s clip, now on 0.1.13 with mlx 0.32.2:

| Model | vocals | drums | bass | guitar | other | piano |
|---|---|---|---|---|---|---|
| `htdemucs_6s` (MLX) | -23.3 | -24.1 | -29.1 | -30.4 | -31.5 | -79.9 |

`htdemucs` in MLX now matches PyTorch demucs on the same clip to within 0.1 dB per stem (vocals -23.1 / -23.1, drums -23.8 / -23.8, bass -29.0 / -29.0, other -27.1 / -27.0). The remaining difference between the two is -20 to -25 dB relative, which is just the random shift offsets differing between the runs. The near-silent piano stem is the clip, not the model — there is no piano in that section.

One note for anyone else upgrading from 0.1.7. `pip install -U` left me with the new `mlx-audio-io` wheel and my old pinned `mlx==0.31.2`, and the new preflight check caught it with a very clear message — that check is a nice addition. After upgrading to `mlx==0.32.2` the prebuilt wheel still failed at import:

```
Symbol not found: __ZN3mlx4core6astypeENS0_5arrayENS0_5DtypeENSt3__17variantIJNS3_9monostateENS0_6StreamENS0_17ThreadLocalStreamENS0_6DeviceEEEE
  Referenced from: mlx_audio_io/_core.cpython-313-darwin.so
  Expected in:     mlx/lib/libmlx.dylib
```

Rebuilding from source against the installed MLX fixed it, exactly as the error message suggested. The only thing the message did not mention is that the build needs CMake and Ninja, which were not in my venv:

```bash
pip install "mlx==0.32.2" "nanobind==2.15.0" scikit-build-core delocate cmake ninja
pip install --force-reinstall --no-cache-dir --no-build-isolation --no-binary mlx-audio-io mlx-audio-io
```

Might be worth a line in the troubleshooting docs, since the rebuild itself takes about ten seconds once the tooling is there.

Good luck with the project — having proper stem separation run on the GPU with no PyTorch at inference time is genuinely useful, and I will be using it on my own tracks.

**Comment by ssmall256** [comment]
Thanks for confirming and sharing the per-stem numbers — matching PyTorch demucs within 0.1 dB is great to see, and good to know the quiet piano stem was just the clip.

Both issues are addressed in mlx-audio-separator 0.1.16 and mlx-audio-io 1.3.19:

⚬ Missing build dependencies: Added cmake and ninja to all error/preflight messages and the README (required when building with --no-build-isolation). Added documentation on why --no-cache-dir is necessary to prevent pip from reusing stale wheels.
⚬ Symbol not found after MLX upgrades: Added a Troubleshooting section to the README explaining why native extensions must be recompiled when MLX changes ABI.

Updated rebuild recipe:

# nanobind must match your MLX: 0.31.x -> 2.12.0, 0.32.x -> 2.15.0
pip install "mlx==0.32.2" "nanobind==2.15.0" scikit-build-core cmake ninja delocate
pip install --force-reinstall --no-cache-dir --no-build-isolation \
  --no-binary mlx-audio-io mlx-audio-io


To update:

pip install -U mlx-audio-separator mlx-audio-io


Thanks again for the detailed report and benchmarks — made this very straightforward to track down; and, I'm glad you find the project useful!
